src/vm_cpu.rs

```diff
@@ -1102,63 +1098,69 @@ pub enum VmCpuStatus {
 struct VmCpusInner {
     // Locking: status must be locked before vcpu.
     status: RwLock<VmCpuStatus>,
-    vcpu: Mutex<VmCpu>,
+    vcpu: Once<PageBox<Mutex<VmCpu>>>,
+}
+
+impl Default for VmCpusInner {
+    fn default() -> Self {
+        VmCpusInner {
+            status: RwLock::new(VmCpuStatus::NotPresent),
+            vcpu: Once::new(),
+        }
+    }
 }
 
 /// The set of vCPUs in a VM.
 pub struct VmCpus {
-    inner: PageVec<VmCpusInner>,
+    num: RwLock<usize>,
+    vcpus: [VmCpusInner; MAX_CPUS],
 }
 
 impl VmCpus {
     /// Creates a new vCPU tracking structure backed by `pages`.
-    pub fn new(
-        guest_id: PageOwnerId,
-        pages: SequentialPages<InternalClean>,
-        page_tracker: PageTracker,
-    ) -> Result<Self> {
-        let num_vcpus = pages.length_bytes() / VM_CPU_BYTES;
-        if num_vcpus == 0 {
-            return Err(Error::InsufficientVmCpuStorage);
-        }
-        let mut inner = PageVec::new(pages, page_tracker);
-        for i in 0..num_vcpus {
-            let entry = VmCpusInner {
-                status: RwLock::new(VmCpuStatus::NotPresent),
-                vcpu: Mutex::new(VmCpu::new(i, guest_id)),
-            };
-            inner.push(entry);
+    pub fn new() -> Self {
+        let vcpus = [(); MAX_CPUS].map(|_| VmCpusInner::default());
+        Self {
+            num: RwLock::new(0),
+            vcpus,
         }
-        Ok(Self { inner })
     }
 
     /// Returns the number of vCPUs in this `VmCpus`.
     pub fn num_vcpus(&self) -> usize {
-        self.inner.len()
+        *self.num.read()
     }
 
     /// Adds the vCPU at `vcpu_id` as an available vCPU using `shared_area` as the vCPU's shared
     /// state-memory state area.
-    pub fn add_vcpu(&self, vcpu_id: u64, shared_area: VmCpuSharedArea) -> Result<()> {
-        let entry = self.inner.get(vcpu_id as usize).ok_or(Error::BadCpuId)?;
+    pub fn add_vcpu(&self, vcpu_id: u64, vcpu_box: PageBox<Mutex<VmCpu>>) -> Result<()> {
+        let entry = self.vcpus.get(vcpu_id as usize).ok_or(Error::BadCpuId)?;
         let mut status = entry.status.write();
         if *status != VmCpuStatus::NotPresent {
             return Err(Error::VmCpuExists);
         }
-        entry.vcpu.lock().shared_area.call_once(|| shared_area);
         *status = VmCpuStatus::PoweredOff;
+
+        assert!(!entry.vcpu.is_completed());
+        entry.vcpu.call_once(|| vcpu_box);
+
+        {
+            let mut n = self.num.write();
```

Comment:
Same type system comment about the length and the storage of VCPUs being held separately. They should be bundled like Vec.